### PR TITLE
IGNITE-21911 Change API usage of Placement driver in Index module from TablePartitionId to ZonePartitionId

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-21911

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java [hunk 1]
```diff
@@ -244,7 +258,7 @@ private void tryScheduleBuildIndex(
      * @param replicaId Replica ID.
      */
     private void stopBuildingIndexesIfPrimaryExpired(TablePartitionId replicaId) {
```
Comment (JAkutenshi) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:260]: The same, shouldn't be there new `ZonePartitionId`?

Comment (JAkutenshi) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:260]: As an additional question: isn't one of subtask is to get rid of `TablePartitionId`?

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/ReplicaAwareLeaseTracker.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            TimeUnit unit
+    ) {
+        TablePartitionId tablePartitionId = new TablePartitionId(groupId.tableId(), groupId.partitionId());
+
+        return delegate.awaitPrimaryReplica(tablePartitionId, timestamp, timeout, unit)
+                .thenCompose(replicaMeta -> {
+                    ClusterNode leaseholderNode = clusterNodeResolver.getById(replicaMeta.getLeaseholderId());
+
+                    WaitReplicaStateMessage awaitReplicaReq = REPLICA_MESSAGES_FACTORY.waitReplicaStateMessage()
+                            .groupId(tablePartitionId)
```
Comment (JAkutenshi) [inline modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/ReplicaAwareLeaseTracker.java:99]: The same, as input we got ZPId as groupId, but have to use TPId

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 3]
```diff
@@ -796,7 +797,7 @@ public class IgniteImpl implements Ignite {
                 catalogManager,
                 metaStorageMgr,
                 indexManager,
-                placementDriverMgr.placementDriver(),
+                new ReplicaAwareLeaseTracker(placementDriverMgr.placementDriver(), replicaSvc, clusterSvc.topologyService()),
```
Comment (JAkutenshi) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:800]: Why we should use the delegate only for index?

**Review by JAkutenshi — COMMENTED** [review]
Mostly questions as a review

File: modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+package org.apache.ignite.internal.replicator;
+
+import java.util.Objects;
+
+/**
+ * The class is used to identify a zone replication group id for a given partition.
+ */
+public class ZonePartitionId implements ReplicationGroupId {
+
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java:26]: Please remove the empty line.

File: modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    /** Zone id. */
+    private final int zoneId;
+
+    /** Partition id. */
+    private final int partId;
+
+    private final int tableId;
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java:33]: Please either add javadoc to the tableId field or remove zoneId, partId ones. From my point of view, the latter is better here.

File: modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * The constructor.
+     *
+     * @param zoneId Zone id.
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java:38]: tableId param is missing.

File: modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+     * @param partId Partition id.
+     */
+    public ZonePartitionId(int zoneId, int partId, int tableId) {
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/replicator/ZonePartitionId.java:41]: Agree. Moreover, I'd actually adjust the fields order in the same manner.

File: modules/index/src/test/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskControllerTest.java [hunk 8]
```diff
@@ -188,11 +189,13 @@ private void setPrimaryReplicaAnotherNode() {
     }
 
     private void setPrimaryReplica(ClusterNode clusterNode) {
-        TablePartitionId groupId = new TablePartitionId(tableId(), 0);
+        ZonePartitionId zonePartId = new ZonePartitionId(0, 0, tableId());
```
Comment (sanpwc) [inline modules/index/src/test/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskControllerTest.java:192]: Here and there. Using 0 is generally not consistent. `tableId` internally wil retrieve id from CatalogTableDescriptor 
```
    public static CatalogTableDescriptor getTableStrict(CatalogService catalogService, String tableName, long timestamp) {
        CatalogTableDescriptor table = catalogService.table(tableName, timestamp);

        assertNotNull(table, "tableName=" + tableName + ", timestamp=" + timestamp);

        return table;
    }
```
Meaning that we should retrieve zoneId from the same descriptor.

File: modules/index/src/test/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskControllerTest.java [hunk 9]
```diff
@@ -188,11 +189,13 @@ private void setPrimaryReplicaAnotherNode() {
...
 
-        ReplicaMeta replicaMeta = newPrimaryReplicaMeta(clusterNode, groupId, HybridTimestamp.MIN_VALUE, HybridTimestamp.MAX_VALUE);
+        TablePartitionId tablePartId = new TablePartitionId(tableId(), 0);
 
-        assertThat(placementDriver.setPrimaryReplicaMeta(0, groupId, completedFuture(replicaMeta)), willCompleteSuccessfully());
+        ReplicaMeta replicaMeta = newPrimaryReplicaMeta(clusterNode, tablePartId, HybridTimestamp.MIN_VALUE, HybridTimestamp.MAX_VALUE);
```
Comment (sanpwc) [inline modules/index/src/test/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskControllerTest.java:196]: Why it's tablePartId in replicaMeta? We should have new ZonePartitionId there.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java [hunk 10]
```diff
@@ -301,8 +302,8 @@ private int indexId(String indexName) {
         return getIndexIdStrict(catalogManager, indexName, clock.nowLong());
     }
 
-    private TablePartitionId replicaId(int partitionId) {
-        return new TablePartitionId(tableId(), partitionId);
+    private ZonePartitionId replicaId(int partitionId) {
+        return new ZonePartitionId(0, partitionId, tableId());
```
Comment (sanpwc) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java:306]: Same as above, not valid to use 0 here as zoneId, please retrieve the identifier from catalog descriptor like `tableId()` does.

File: modules/index/src/test/java/org/apache/ignite/internal/index/TestPlacementDriver.java [hunk 11]
```diff
@@ -70,7 +71,7 @@ CompletableFuture<Void> setPrimaryReplicaMeta(
                 PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED,
                 new PrimaryReplicaEventParameters(
                         causalityToken,
-                        replicaId,
+                        new TablePartitionId(replicaId.tableId(), replicaId.partitionId()),
```
Comment (sanpwc) [inline modules/index/src/test/java/org/apache/ignite/internal/index/TestPlacementDriver.java:74]: We should use ZonePartitionId here.

File: modules/placement-driver-api/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriver.java [hunk 12]
```diff
@@ -60,6 +61,28 @@ CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
             TimeUnit unit
     );
 
+    /**
+     * Temporary solution for awaiting {@link ReplicaMeta}. Waits for
```
Comment (sanpwc) [inline modules/placement-driver-api/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriver.java:65]: If it's temporary please add todo with a link and mark the method as @Deprecated. Please create a ticket if necessary.

File: modules/placement-driver-api/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriver.java [hunk 13]
```diff
@@ -60,6 +61,28 @@ CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
...
+     * @param groupId Replication group id.
+     * @param timestamp CLOCK_SKEW aware timestamp reference value.
+     * @param timeout How long to wait before completing exceptionally with a TimeoutException, in units of unit.
+     * @param unit A TimeUnit determining how to interpret the timeout parameter.
+     * @return Primary replica future.
+     * @throws PrimaryReplicaAwaitTimeoutException If primary replica await timed out.
+     * @throws PrimaryReplicaAwaitException If primary replica await failed with any other reason except timeout.
+     */
+    default CompletableFuture<ReplicaMeta> awaitPrimaryReplicaForTable(
+            ZonePartitionId groupId,
```
Comment (sanpwc) [inline modules/placement-driver-api/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriver.java:78]: Why it's ZonePartitionId? It should be `ReplicationGroupId`.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java [hunk 14]
```diff
@@ -240,6 +247,18 @@ private CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessag
...
     }
 
+    /**
+     * Process {@link WaitReplicaStateMessage}.
+     *
+     * @param msg Message to process.
+     * @return Future that contains a result.
+     */
+    private CompletableFuture<Void> processWaitReplicaStateMessage(WaitReplicaStateMessage msg) {
+        LOG.info("Received LeaseGrantedMessage for replica belonging to group=" + groupId());
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java:257]: It's not LeaseGrantedMessage.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java [hunk 15]
```diff
@@ -240,6 +247,18 @@ private CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessag
...
+
+        return waitForActualState(FastTimestamps.coarseCurrentTimeMillis() + TimeUnit.SECONDS.toMillis(msg.timeout()));
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java:259]: It's also required to sendPrimaryReplicaChangeToReplicationGroup after waitForActualState. Please check `Replica#processLeaseGrantedMessage` for more details.

**Pull request closed without merging**